### PR TITLE
Expose head service ServiceSpec

### LINKED ISSUE #877: [Feature] Expose loadBalancerIP for head service to users
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

Users may want Ray head to be accessible from outside Kubernetes. One option is to use Kubernetes Service with type `LoadBalancer` and specify `loadBalancerIP`. However, KubeRay currently does not expose `loadBalancerIP` to users. See https://discuss.ray.io/t/specify-ip-adress-for-the-head-as-a-loadbalancer-service-type/9008 for more details.

### Use case

_No response_

### Related issues

https://discuss.ray.io/t/specify-ip-adress-for-the-head-as-a-loadbalancer-service-type/9008

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by architkulkarni**
Exposes the `ServiceSpec` of the head pod service to the user.  Previously, we only exposed the `type` field. This led to limitations such as, for example, the user specifying `type: LoadBalancer` and having no way to set the associated field `LoadBalancerIPs`.

After this PR, we will have two ways of specifying the service type: the legacy `HeadGroupSpec.ServiceType`, and the new `HeadGroupSpec.HeadServiceSpec.Type`.  In this PR we make the former take precedence.

In the long term we should have only one way of specifying this.  One plan might involve a few PRs:
1. Update all sample YAMLS and documentation examples that currently use `HeadGroupSpec.ServiceType` to use `HeadGroupSpec.HeadServiceSpec.Type`, if any
2. Print a deprecation warning for `HeadGroupSpec.ServiceType` in v0.6
3. Remove `HeadGroupSpec.ServiceType` in v0.7.

Or, if we think it's convenient and harmless, we can leave them both in forever.
## Why are these changes needed?
Closes #877 

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
Is https://github.com/ray-project/kuberay/blob/master/ray-operator/controllers/ray/raycluster_controller_fake_test.go the best place to add a test for this?

**Comment by architkulkarni** [comment]
There's one test failure:
```
                               Diff:
                                --- Expected
                                +++ Actual
                                @@ -8,3 +8,5 @@
                                   (string) (len=16) "rayClusterConfig": (map[string]interface {}) (len=3) {
                                -   (string) (len=13) "headGroupSpec": (map[string]interface {}) (len=3) {
                                +   (string) (len=13) "headGroupSpec": (map[string]interface {}) (len=4) {
                                +    (string) (len=15) "headServiceSpec": (map[string]interface {}) {
                                +    },
                                     (string) (len=14) "rayStartParams": (map[string]interface {}) (len=7) {
                Test:           TestMarshallingRayService
--- FAIL: TestMarshallingRayService (0.00s)
```

I'm trying to figure out why the new `headServiceSpec` field is being included but the existing `serviceType` field isn't included, even though they're both optional and have similar code handling them.

File: ray-operator/controllers/ray/common/service.go [hunk 1]
```diff
@@ -40,25 +40,37 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
...
-			Type:     cluster.Spec.HeadGroupSpec.ServiceType,
-		},
-	}
-
-	ports := getServicePorts(cluster)
-	defaultAppProtocol := DefaultServiceAppProtocol
-	for name, port := range ports {
-		svcPort := corev1.ServicePort{Name: name, Port: port, AppProtocol: &defaultAppProtocol}
-		service.Spec.Ports = append(service.Spec.Ports, svcPort)
+		Spec: serviceSpec,
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service.go:73]: `v1.ServiceSpec` is not enough for users. For example, users need to configure head service's annotations for some ingress controllers (#708). In addition, users also want to configure service's labels (#874, [slack thread](https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1673647485229799)).

**Comment by DmitriGekhtman** [comment]
> In the long term we should have only one way of specifying this.

At some point, the API needs to be upgraded to, say, v1beta1. Then you're allowed to make backwards-incompatible changes, but there's some rigamarole with converting webhooks that you're supposed to do: https://book.kubebuilder.io/multiversion-tutorial/conversion-concepts.html

File: helm-chart/ray-cluster/templates/raycluster-cluster.yaml [hunk 2]
```diff
@@ -19,6 +19,9 @@ spec:
   {{- end }}
   headGroupSpec:
     serviceType: {{ .Values.service.type }}
+    {{- if .Values.service.headServiceSpec }}
```
[the remark this answers is not in the document]
Comment (architkulkarni) [reply helm-chart/ray-cluster/templates/raycluster-cluster.yaml:22]: I was wondering about this, why does it need to be included?  I used [head `autoscalerOptions:` ](https://github.com/ray-project/kuberay/blob/master/helm-chart/ray-cluster/values.yaml#L29)as an example, which is also not included in `values.yaml`.  

Did you mean to add `headServiceSpec` with a default value of `{}`, or to add it commented out like `autoscalerOptions`?

File: ray-operator/controllers/ray/common/service.go (hunk 1 above)
Comment (architkulkarni) [reply ray-operator/controllers/ray/common/service.go:73]: Resolved offline. I'll update this PR to expose the entire head service (which in particular includes ObjectMeta, which includes labels and annotations.)

**Comment by architkulkarni** [comment]
Closing in favor of https://github.com/ray-project/kuberay/pull/1040

**Pull request closed without merging**